MediaPlayerPrivateMediaStreamAVFObjC needs to stop double-hosting layers

pvollan opened the pull request:
#### ec2913e079230d38595455a183367ed339e14592
<pre>
MediaPlayerPrivateMediaStreamAVFObjC needs to stop double-hosting layers
<a href="https://bugs.webkit.org/show_bug.cgi?id=253761">https://bugs.webkit.org/show_bug.cgi?id=253761</a>
rdar://106576224

Reviewed by Youenn Fablet and Brent Fulgham.

MediaPlayerPrivateMediaStreamAVFObjC needs to stop double-hosting layers to allow for CARenderServer blocking.
This patch overrides layerHostingContextID() for MediaPlayerPrivateMediaStreamAVFObjC, which will use the
hosting context ID from SampleBufferDisplayLayer. SampleBufferDisplayLayer receives this context ID from the
GPU process. Additionally, setVideoInlineSizeFenced() is overridden to set the size of the
SampleBufferDisplayLayer. When EWS was processing this patch, I saw timeouts of the test
fast/mediastream/getUserMedia-to-canvas-2.html on iOS simulator. I have not been able to reproduce this
locally. I marked the test as flaky, and filed <a href="https://bugs.webkit.org/show_bug.cgi?id=253959">https://bugs.webkit.org/show_bug.cgi?id=253959</a> to investigate
this further.

LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/platform/graphics/avfoundation/SampleBufferDisplayLayer.h:
(WebCore::SampleBufferDisplayLayer::hostingContextID const):
* Source/WebCore/platform/graphics/avfoundation/SampleBufferDisplayLayer.h:
* Source/WebCore/platform/graphics/avfoundation/objc/LocalSampleBufferDisplayLayer.h:
* Source/WebCore/platform/graphics/avfoundation/objc/LocalSampleBufferDisplayLayer.mm:
(WebCore::LocalSampleBufferDisplayLayer::updateBoundsAndPosition):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::hostingContextID const):
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::setVideoInlineSizeFenced):
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.cpp:
(WebKit::RemoteSampleBufferDisplayLayer::updateBoundsAndPosition):
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.h:
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.messages.in:
* Source/WebKit/WebProcess/GPU/webrtc/SampleBufferDisplayLayer.cpp:
(WebKit::SampleBufferDisplayLayer::updateBoundsAndPosition):
* Source/WebKit/WebProcess/GPU/webrtc/SampleBufferDisplayLayer.h:

Canonical link: <a href="https://commits.webkit.org/261709@main">https://commits.webkit.org/261709@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc59f21aa0ba5ae0f4b5031069c27c261022cc0a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112617 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21771 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1282 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4394 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/121158 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116684 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23111 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12936 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/5546 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118385 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17165 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100388 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105690 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/99118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/88/builds/934 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/46190 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14104 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/973 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/95294 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14782 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20122 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52968 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8166 "'git push ...'") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16624 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.h:
```diff
@@ -231,6 +231,9 @@ class MediaPlayerPrivateMediaStreamAVFObjC final
     MediaStreamTrackPrivate* activeVideoTrack() const;
 
+    LayerHostingContextID hostingContextID() const override;
+    void setVideoInlineSizeFenced(const FloatSize&, const WTF::MachSendRight&) override;
```
We can make them final.
Ideally we would move them to the SampleBufferDisplayLayer::Client section. Strangely, this section is currently below, it could probably be changed.

youennf commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.mm:
```diff
@@ -1157,6 +1157,22 @@ void updateTracksOfKind(MemoryCompactRobinHoodHashMap<String, RefT>& trackMap, T
+{
+    if (m_sampleBufferDisplayLayer)
+        return m_sampleBufferDisplayLayer->hostingContextID();
+    return { };
```
oneliner?
return m_sampleBufferDisplayLayer ? m_sampleBufferDisplayLayer->hostingContextID(); : 0;

youennf commented on Source/WebKit/WebProcess/GPU/webrtc/SampleBufferDisplayLayer.h:
```diff
@@ -51,6 +51,8 @@ class SampleBufferDisplayLayer final : public WebCore::SampleBufferDisplayLayer,
 
     void didReceiveMessage(IPC::Connection&, IPC::Decoder&) final;
 
+    LayerHostingContextID hostingContextID() const override { return m_hostingContextID; }
```
final

youennf commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.mm:
```diff
@@ -1157,6 +1157,22 @@ void updateTracksOfKind(MemoryCompactRobinHoodHashMap<String, RefT>& trackMap, T
+    return { };
+}
+
+void MediaPlayerPrivateMediaStreamAVFObjC::setVideoInlineSizeFenced(const FloatSize& size, const WTF::MachSendRight&)
```
I am not familiar with setVideoInlineSizeFenced. I see this is related to full screen.
Why aren't we using the second parameter like for the remote player code path, for instance to RemoteSampleBufferDisplayLayer::m_layerHostingContext?

pvollan replied:
> I am not familiar with setVideoInlineSizeFenced. I see this is related to full screen. Why aren't we using the second parameter like for the remote player code path, for instance to RemoteSampleBufferDisplayLayer::m_layerHostingContext?

That is a good point. I have addressed this in the latest patch by sending the fence back to the GPU process.

Thanks for reviewing!

brentfulgham reviewed (approved):
Looks good, and thank you for forward-declaring LayerHostingContextID instead of including the header.

youennf commented on Source/WebCore/platform/graphics/avfoundation/SampleBufferDisplayLayer.h:
```diff
@@ -62,7 +65,7 @@ class SampleBufferDisplayLayer {
 
     virtual void updateAffineTransform(CGAffineTransform) = 0;
-    virtual void updateBoundsAndPosition(CGRect, VideoFrameRotation) = 0;
+    virtual void updateBoundsAndPosition(CGRect, VideoFrameRotation, std::optional<WTF::MachSendRight> = std::nullopt) = 0;
```
Should probably be std::optional<WTF::MachSendRight>&&, given MachSendRight has a move constructor

pvollan replied:
Good point

> Looks good, and thank you for forward-declaring LayerHostingContextID instead of including the header.

Thanks for reviewing!

pvollan replied:
> Should probably be std::optionalWTF::MachSendRight&&, given MachSendRight has a move constructor

Good point, will fix!

youennf commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.h:
```diff
@@ -231,6 +231,12 @@ class MediaPlayerPrivateMediaStreamAVFObjC final
 
+    // SampleBufferDisplayLayer::Client
+    void sampleBufferDisplayLayerStatusDidFail() final;
+
```
I would remove this line

pvollan replied:
The new virtual methods are not declared in SampleBufferDisplayLayer::Client, so I wonder if it makes sense to keep the extra line here. What do you think?

brentfulgham replied:
Why did you need to move this declaration? Did you get a compile error? If so, I think you should leave your change as-is.

brentfulgham reviewed (approved):
r=me

pvollan committed:
MediaPlayerPrivateMediaStreamAVFObjC needs to stop double-hosting layers
https://bugs.webkit.org/show_bug.cgi?id=253761
rdar://106576224

Reviewed by Youenn Fablet and Brent Fulgham.

MediaPlayerPrivateMediaStreamAVFObjC needs to stop double-hosting layers to allow for CARenderServer blocking.
This patch overrides layerHostingContextID() for MediaPlayerPrivateMediaStreamAVFObjC, which will use the
hosting context ID from SampleBufferDisplayLayer. SampleBufferDisplayLayer receives this context ID from the
GPU process. Additionally, setVideoInlineSizeFenced() is overridden to set the size of the
SampleBufferDisplayLayer. When EWS was processing this patch, I saw timeouts of the test
fast/mediastream/getUserMedia-to-canvas-2.html on iOS simulator. I have not been able to reproduce this
locally. I marked the test as flaky, and filed https://bugs.webkit.org/show_bug.cgi?id=253959 to investigate
this further.

LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/platform/graphics/avfoundation/SampleBufferDisplayLayer.h:
(WebCore::SampleBufferDisplayLayer::hostingContextID const):
* Source/WebCore/platform/graphics/avfoundation/SampleBufferDisplayLayer.h:
* Source/WebCore/platform/graphics/avfoundation/objc/LocalSampleBufferDisplayLayer.h:
* Source/WebCore/platform/graphics/avfoundation/objc/LocalSampleBufferDisplayLayer.mm:
(WebCore::LocalSampleBufferDisplayLayer::updateBoundsAndPosition):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::hostingContextID const):
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::setVideoInlineSizeFenced):
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.cpp:
(WebKit::RemoteSampleBufferDisplayLayer::updateBoundsAndPosition):
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.h:
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.messages.in:
* Source/WebKit/WebProcess/GPU/webrtc/SampleBufferDisplayLayer.cpp:
(WebKit::SampleBufferDisplayLayer::updateBoundsAndPosition):
* Source/WebKit/WebProcess/GPU/webrtc/SampleBufferDisplayLayer.h:

Canonical link: https://commits.webkit.org/261709@main
